### PR TITLE
Inline sync plugin

### PR DESCRIPTION
**Pull request by Saul-Mirone**
-   [x] I read the contributing guide 
-   [x] I agree to follow the code of conduct 

## Summary

There're lots of bugs caused by consistency between markdown AST and UI.

For example:
https://github.com/Saul-Mirone/milkdown/issues/637
https://github.com/Saul-Mirone/milkdown/issues/553
https://github.com/Saul-Mirone/milkdown/issues/622

The `inline-sync-plugin` can help to solve this issue. Instead of using inputrules to check user input, this plugin keep sync the prosemirror node between markdown AST for inline nodes (such as `paragraph` and `heading`).

## How did you test this change?

Manually test and e2e test.

### CODE REVIEW

**Comment by 4-1-1** [comment]
Seemly , It will broken the editor history.

**Comment by Saul-Mirone** [comment]
> Seemly , It will broken the editor history.

Do you have reproducable steps of videos?

**Comment by 4-1-1** [comment]
> > Seemly , It will broken the editor history.
> 
> Do you have reproducable steps of videos?

#704 like this

**Comment by 4-1-1** [comment]
Could you make the Inline sync plugin  as an option. because I have do support  `abc**de**` style .
by override a package name "micromark-util-classify-character"

**Comment by Saul-Mirone** [comment]
> Could you make the Inline sync plugin as an option. because I have do support `abc**de**` style . by override a package name "micromark-util-classify-character"

I can certiainly make it can be turned off, but in my opinion `abc**de**` can work as usual.
https://milkdown.dev/online-demo?text=IYIwxgVBAmCmUCgg

**Comment by 4-1-1** [comment]
> > Could you make the Inline sync plugin as an option. because I have do support `abc**de**` style . by override a package name "micromark-util-classify-character"
> 
> I can certiainly make it can be turned off, but in my opinion `abc**de**` can work as usual. https://milkdown.dev/online-demo?text=IYIwxgVBAmCmUCgg

sorry , `abc**'de**`abc**'de**

**Comment by Saul-Mirone** [comment]
> > > Could you make the Inline sync plugin as an option. because I have do support `abc**de**` style . by override a package name "micromark-util-classify-character"
> > 
> > 
> > I can certiainly make it can be turned off, but in my opinion `abc**de**` can work as usual. https://milkdown.dev/online-demo?text=IYIwxgVBAmCmUCgg
> 
> sorry , `abc**'de**`abc**'de**

Do you mean you need `abd**'de**` to render the `'de` as a strong mark? I think that's not allowed by markdown and only possible if you make a remark plugin to support that. The previous render behavior of milkdown is a bug.

**Comment by 4-1-1** [comment]
I have done it. but maybe will bring some other bugs,not sure, runs fine right now

**Comment by z-950** [comment]
Early parsing markdown break plugins that uses `remark-directive`.
Because `remark-directive` has no name whitelist. It will throw error `Cannot match target parser` at the first character after `:` and then can not input but delete `:`.
So any plugins that like this is also broken.

Strictly speaking, `remark-directive` must have name whitelist, because it will match `:name` as text (also leaf, container) directive. `:name` is so common to appear in content.

**Comment by 4-1-1** [comment]
@z-950 look at this [@ezone-devops/milkdown-plugin-directive-fallback](https://www.npmjs.com/package/@ezone-devops/milkdown-plugin-directive-fallback)

**Comment by z-950** [comment]
@4-1-1 It's not a perfect realization of the fallback. I try do this, but it will lost some others directive.
This plugin can't solve nested directives. And some other lost like: `:i[lovely]` becomes `:i`.

**Comment by z-950** [comment]
@4-1-1 
The whitelist of remark-directive is the only one solution.
So the current fallback is like this: writing our `remark-directive`, which means to control the markdown ast result.
Only need change `factory-name` in `micromark-extension-directive` imported by `remark-directive`.
```js
export function factoryName(effects, ok, nok, type) {
  const self = this;

  let nameCache = ''; // {1}

  return start;

  /** @type {State} */
  function start(code) {
    if (asciiAlpha(code)) {
      nameCache = String.fromCharCode(code); // {2}
      effects.enter(type);
      effects.consume(code);
      return name;
    }

    return nok(code);
  }

  /** @type {State} */
  function name(code) {
    if (
      code === codes.dash ||
      code === codes.underscore ||
      asciiAlphanumeric(code)
    ) {
      nameCache += String.fromCharCode(code); // {3}
      effects.consume(code);
      return name;
    }

    effects.exit(type);
    return self.previous === codes.dash || self.previous === codes.underscore || !whitelist.includes(nameCache) //{4}
      ? nok(code)
      : ok(code);
  }
}
```

**Comment by Saul-Mirone** [comment]
I've added some options for sync plugin to make the behavior can be controled by users.

File: packages/preset-commonmark/src/mark/strong.ts [hunk 1]
```diff
@@ -13,6 +12,7 @@ export const strong = createMark<Keys>((utils) => {
     return {
         id,
         schema: () => ({
+            inclusive: false,
```
Comment (iHaPBoy) [inline packages/preset-commonmark/src/mark/strong.ts:15]: `inclusive: false` causes the mark to end after one character is entered.

inclusive

Comment (Saul-Mirone) [reply packages/preset-commonmark/src/mark/strong.ts:15]: It's a trade off to support inline sync. We need to provide custom toggle mark command to fix this.

Comment (iHaPBoy) [reply packages/preset-commonmark/src/mark/strong.ts:15]: @Saul-Mirone Okay, I've created an issue for this. #809

**Pull request merged**